Add wazero specific limitations (#186)
tetratelabs/wazero: RATIONALE.md

https://github.com/bytecodealliance/wasmtime/blob/v0.29.0/crates/lightbeam/src/microwasm.rs

## Implementation limitations

> [WebAssembly specification states that](https://www.w3.org/TR/wasm-core-1/#a2-implementation-limitations)

Reviewer: WebAssembly 1.0 (MVP) specification allows runtimes to limit certain aspects of Wasm module or execution.

runtime implementors can impose their own limits on a number of aspects of a Wasm module or execution.

> The followings are the limitations we explicitly impose and handle as errors during module instantiation in wazero:

Reviewer: wazero limitations are imposed pragmatically and described below.

### Number of functions in a store

The possible number of function instances in [a store](https://www.w3.org/TR/wasm-core-1/#store%E2%91%A0) is not specified in the WebAssembly specifications since [`funcaddr`](https://www.w3.org/TR/wasm-core-1/#syntax-funcaddr) corresponding to a function instance can be arbitrary number. 

> In wazero, we choose to set the maximum number of function instances a store can instantiate is limited to 2^27 as that would result in occupying 2^20 (= 1 GiB) bytes in memory for function pointers alone.

Reviewer: wazero limits the maximum function instances to 2^27 as even that number would occupy 1GB in function pointers

That is because not only we _believe_ that all use cases are fine with the limitation, but also we have no way to test wazero runtimes under these unusual circumstances.

### Number of function types in a store

There's no limitation on the number of function types in [a store](https://www.w3.org/TR/wasm-core-1/#store%E2%91%A0) according to the spec. In wazero implementation, we assign each function type to a unique ID, and choose to use `uint32` to represent the IDs.
Therefore the maximum number of function types a store can have is limited to 2^27. 

This is due to the same reason for the limitation on the number of functions above.

> ### Number of values on the stack in a function

Author: handling this limit as an error would be a bit large change so I will do in a followup PR

> Therefore the maximum number of function types a store can have is limited to 2^27 as that would result in occupying 2^29 = (512 MiB) bytes in memory for function types alone.

Reviewer: to 2^27 as even that number would occupy 512MB just to reference the function types.

> According to the spec, there's no limitation on the number of values we a function can retain in the Wasm values stack. We limit the maximum number to 2^27 = 134,217,728.

Reviewer: While the the spec does not clarify a limitation of function stack values, wazero limits this to 2^27.

> The reason is that we internally represent all the values as 64-bit integes regardless of its types (including f32, f64), and 2^27 values means 

Reviewer: ```suggestion
The reason is that we internally represent all the values as 64-bit integers regardless of its types (including f32, f64), and 2^27 values means 
```

1 GiB = (2^30). 1 GiB is the reasonable for most applications [as we see a Goroutine has 250 MB as a limit on the stack for 32-bit arch](https://github.com/golang/go/blob/f296b7a6f045325a230f77e9bda1470b1270f817/src/runtime/proc.go#L120), considering that WebAssembly is (currently) 32-bit environment.

> All the functions are statically analyzed at module insntantiation phase, and if a function can potentially reach this limit, an error is returend.

Reviewer 2: Found some typos.
```suggestion
All the functions are statically analyzed at module instantiation phase, and if a function can potentially reach this limit, an error is returned.
```

### Number of globals in a module

> Theoretically, a module can declare globals (inclding imported ones) up to 2^32 times. However, we limit the number of available globals in a module to 2^27 = 134,217,728.

Reviewer: ```suggestion
Theoretically, a module can declare globals (including imports) up to 2^32 times. However, wazero limits this to  2^27(134,217,728) per module.
```
